### ISSUE TITLE
Uploads to Amazon Cloud Drive restarting at 99%

### ISSUE THREAD

**Issue by zod-melodic** [ncw/rclone#601]
Using rclone 1.32 on Windows Server 2012 R2
Connecting to Amazon Cloud Drive
Command used:` rclone sync I:\MOVIES ACD:SERVERNAME\MOVIES --bwlimit 24M --max-size=50G --transfers=6 --exclude ".*/" `

For the past ten days to two weeks, many of my uploads to Amazon hit 99% and then restart at 0%. When I run the command with the verbose command, I regularly see one of the following error messages (in order of frequency):

`error Post https://content-na.drive.amazonaws.com/cdproxy/nodes?suppress=deduplication: write tcp 192.168.2.99:52277->xx.xxx.xxx.xxx:443: wsasend: An existing connection was forcibly closed by the remote host`

`Received error: HTTP code 504: "504 GATEWAY_TIMEOUT", no response body` `- low level retry`
(these low level retries sometimes get up to four or five before finally succeeding)

`Received error: HTTP code 500: "500 Internal Server Error", reponse body: {"message":"Internal failure"}
`

Is there something happening on Amazon's end? I upped the transfer number to six in the hopes that more attempts would lead to more successful uploads and it seems to be working, but I am still seeing regular resets.

**Comment by ncw** [comment]
I think this is a problem at Amazon's end...

If you check out the [developer forum](https://forums.developer.amazon.com/spaces/87/Cloud+Drive.html) you'll see various relevant threads ([one by me](https://forums.developer.amazon.com/questions/22563/408-request-timeout-errors.htm)!).

**Comment by zod-melodic** [comment]
Yeah, I guess so. The problem has definitely gotten worse lately. Thanks for pointing me to the forum.

**Comment by goldelehr** [comment]
@ncw Looks like your thread was deleted. Amazon mentions a new API in their forums

https://forums.developer.amazon.com/questions/33186/the-new-amazon-drive-api-and-sdks-for-android-and-1.html

maybe this will fix the problem with files larger than 10 GB?

**Comment by ncw** [comment]
[Here is the thread](https://forums.developer.amazon.com/questions/22563/408-request-timeout-errors.html) I missed the last character off the URL when pasting it!

I can't decide if it is a new API or just a wrap of the existing API for android & IOS but I sent off a request for an invite and I'll let you know!

**Comment by ncw** [comment]
I posted a potential fix for this in #606 if you fancy giving it a go.

**Comment by zod-melodic** [comment]
I'll try this over the weekend and see how it works!

**Comment by zod-melodic** [comment]
Sorry, but would it be possible to build a version for Windows x64? My home server runs Windows Server 2012 R2.

**Comment by ncw** [comment]
See #606

**Comment by ncw** [comment]
Note to self - maybe combine this fix with #559

**Comment by zod-melodic** [comment]
Looks like the problems on Amazon's end are getting worse. From looking at the log, there have been several 504 errors mid-transfer, as opposed to at the end. 

The patch seems to have mostly worked on the occasion where the error came at the end. It waited, then looked again and found the file with the correct size and indicated that it was returning with no error. However then it threw this error

'Received error: %!v(PANIC=runtime error: invalid memory address or nil pointer dereference)'

And started the file all over again.

I'll upload the entire log on Monday so that there's an entire weekend's worth of data to pore through, but as of this point I've uploaded 175 GB worth of data and managed to successfully complete one (1) file. Something is very wrong on Amazon's end.

**Comment by zod-melodic** [comment]
I think there might be a bug in the patch.

Here's what happens after any file reaches 100% but then gets a timeout error:

```
Timeout error detected - waiting to see if object was uploaded correctly: HTTP code 504: "504 GATEWAY_TIMEOUT", no response body
Object found with correct size - returning with no error
pacer: Rate limited, sleeping for 150.761883ms (1 consecutive low level retries)
pacer: low level retry 1/1 (error <nil>)
Received error: %!v(PANIC=runtime error: invalid memory address or nil pointer dereference) - low level retry 2/10
Removing failed copy
Resetting sleep to minimum 20ms on success
```

I think it's seeing that it's uploaded correctly and is the correct size, but then it throws the PANIC=runtime error and decides it's a bad copy, deletes it, and starts over from the beginning again.

**Comment by zod-melodic** [comment]
Here's the full log from this weekend's attempt. 
[test.txt](https://github.com/ncw/rclone/files/407133/test.txt)

**Comment by ncw** [comment]
Very useful log, thanks.

I can see that if I hadn't made a mistake this would have worked 15 times and not worked 23 times which is not bad.

The time for a file to appear after the TIMEOUT error was from 0-15 seconds which is useful information.

At `2016/08/06 04:04:3` something interesting happened.  Your token expired and at the same time all 6 of your uploads died, none of them near 100%.  Then for 10 minutes or so ACD timedout all your uploads - in fact most of the errors seemed to happen about here.

I see what happened there - because all 6 of the upload threads were busy, the token didn't get renewed in a timely fashion. The previous token renewal was at `2016/08/06 01:55:51` so 2 hours and 10 minutes before. The tokens normally only last for one hour.

I'll have a go at fixing that too and that should make things much more reliable.

I've fixed the mistake in the original patch (hopefully!) - here is a new one with the fix for the token timeout also (not very well tested!).

http://pub.rclone.org/v1.32-18-gaffc483%CE%B2/

**Comment by ncw** [comment]
OK here is another attempt, a bit better tested:

http://pub.rclone.org/v1.32-18-gaffc483%CE%B2/

**Comment by zod-melodic** [comment]
I'll give this latest build a spin throughout the week and let you know the results!

**Comment by ncw** [comment]
With correct URL this time: http://pub.rclone.org/v1.32-18-g29ab7ea%CE%B2/

**Comment by ncw** [comment]
I've reworked the patch again with the following changes.  On a failed upload we
  * detect if we uploaded the file to 100% - if we didn't do nothing
  * do this for any upload error (so might get round the 500 error)
  * increase timeout to 2 minutes - timeout might need to vary with file size?
  * made timeout configurable with `--acd-upload-wait-time` - perhaps set it large when uploading huge files?

Here is the beta: http://pub.rclone.org/v1.32-18-gf7562dc%CE%B2/

**Comment by ser** [comment]
Hi, what do you mean by 'large', I have ~25G file, what should I specify? :)

**Comment by ncw** [comment]
Sorry missed your message - you could try `--acd-upload-wait-time 10m` to wait 10 minutes. I don't know the optimum size though.

**Comment by ncw** [comment]
I've merged this to master and it will be in the v1.33 release - thanks for your help testing.

Please re-open if you see any problems.

Here is a beta with the latest code in: http://pub.rclone.org/v1.32-20-g2ebeed6%CE%B2/

**Comment by zod-melodic** [comment]
Thanks again for addressing this issue so promptly and effectively!
